ibm_i/datadog_checks/ibm_i/check.py

```diff
@@ -78,17 +69,125 @@ def check(self, _):
+        self._delete_connection_subprocess()
+
+    def handle_query_error(self, error):
+        self._current_errors += 1
+        return error
+
+    def _create_connection_subprocess(self):
+        self._subprocess = subprocess.Popen(
+            # TODO: Change to sys.executable once we're based on Agent 7.30.0
+            [
+                "/opt/datadog-agent/embedded/bin/python3",
+                "-c",
+                "from datadog_checks.ibm_i.query_script import query; query()",
+            ],
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            text=True,
+        )
+
+        # Set stdout reader as non-blocking, we don't want to
+        # block .read() calls to be able to time out.
+        fl = fcntl.fcntl(self._subprocess.stdout.fileno(), fcntl.F_GETFL)
+        fcntl.fcntl(self._subprocess.stdout, fcntl.F_SETFL, fl | os.O_NONBLOCK)
+
+        # Set stderr reader as non-blocking, we don't want to
+        # wait until EOF is sent, we only want to read whatever is there when
+        # we try to return errors.
+        fl = fcntl.fcntl(self._subprocess.stderr.fileno(), fcntl.F_GETFL)
+        fcntl.fcntl(self._subprocess.stderr, fcntl.F_SETFL, fl | os.O_NONBLOCK)
+
+        try:
+            print(self.connection_string, file=self._subprocess.stdin, flush=True)
+        except BrokenPipeError as e:
+            # The stdin pipe is broken, usually due to the Agent
+            # killing the subprocess when stopping.
+            # Clean up then return.
+            self._delete_connection_subprocess(e)
+
+    def _delete_connection_subprocess(self, message):
+        if self._subprocess:
+            self.log.debug("Deleting connection: %s", message)
+            while not self._subprocess.returncode:
+                self._subprocess.kill()
+                self._subprocess.wait()
+
+        self._subprocess = None
+
+    def execute_query(self, query, disconnect_on_error=True):
+        if not self._subprocess:
+            self._create_connection_subprocess()
 
-            # https://github.com/mkleehammer/pyodbc/wiki/Cursor
-            for row in cursor:
-                yield row
+        try:
+            # Write query
+            print(query, file=self._subprocess.stdin, flush=True)
+        except BrokenPipeError as e:
+            # The stdin pipe is broken, usually due to the Agent
+            # killing the subprocess when stopping.
+            # Clean up then return.
+            self._delete_connection_subprocess(e)
+            return
+
+        done = False
+        query_start = datetime.now()
+
+        while not done and (datetime.now() - query_start).total_seconds() <= self.config.query_timeout:
+            # Sleep for a bit to wait for results & avoid being a busy loop
+            time.sleep(0.1)
+            try:
+                lines = self._subprocess.stdout.read().strip().split(os.linesep)
+                for line in lines:
```

Comment:
```suggestion
                # To avoid blocking never use a pipe's file descriptor iterator. See https://bugs.python.org/issue3907
                for line in iter(self.connection_subprocess.stdout.readline, ''):
```